cli/src/parse.rs

```diff
@@ -393,6 +526,177 @@ pub fn parse_file_at_path(parser: &mut Parser, opts: &ParseFileOptions) -> Resul
+            if line.is_empty() {
+                break;
+            }
+            if !opts.no_ranges {
+                write!(
+                    stdout,
+                    "\n{}{}{quote}{}{quote}",
+                    render_node_range(opts, cursor, row_width),
+                    "  ".repeat(indent_level + 1),
+                    &paint(color, &render_node_text(line)),
+                )?;
+            } else {
+                write!(
+                    stdout,
+                    "\n{}{quote}{}{quote}",
+                    "  ".repeat(indent_level + 1),
+                    &paint(color, &render_node_text(line)),
+                )?;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn render_node_range(opts: &ParseFileOptions, cursor: &TreeCursor, row_width: usize) -> String {
+    let node = cursor.node();
+    let is_named = node.is_named();
+    let start = node.start_position();
+    let end = node.end_position();
+    let range_color = if is_named && cursor.field_name().is_none() {
+        opts.parse_theme.row_color_named
+    } else {
+        opts.parse_theme.row_color
+    };
+    paint(
+        range_color,
+        &format!(
+            "{:row_width$}:{:<3} - {:row_width$}:{:<3}",
+            start.row, start.column, end.row, end.column,
+        ),
+    )
+}
+
+fn cst_render_node(
+    opts: &ParseFileOptions,
+    cursor: &mut TreeCursor,
+    source_code: &[u8],
+    stdout: &mut StdoutLock<'static>,
+    row_width: usize,
+    indent_level: usize,
+) -> Result<()> {
+    let node = cursor.node();
+    let is_named = node.is_named();
+    if !opts.no_ranges {
+        write!(stdout, "{}", render_node_range(opts, cursor, row_width))?;
+    }
+    write!(stdout, "{}", "  ".repeat(indent_level))?;
+    if is_named {
+        if let Some(field_name) = cursor.field_name() {
+            write!(
+                stdout,
+                "{}",
+                paint(opts.parse_theme.field, &format!("{field_name}: "))
+            )?;
+        }
+        let kind_color = if node.is_error() {
+            opts.parse_theme.error
+        } else if node.is_extra() || node.parent().is_some_and(|p| p.is_extra()) {
+            opts.parse_theme.extra
+        } else {
+            opts.parse_theme.node_kind
+        };
+        write!(stdout, "{} ", paint(kind_color, node.kind()))?;
+        if node.child_count() == 0 {
+            let color = if source_code.is_utf8() {
+                opts.parse_theme.node_text
+            } else {
+                opts.parse_theme.error
+            };
+            write_node_text(
+                opts,
+                stdout,
+                cursor,
+                &String::from_utf8_lossy(&source_code[node.start_byte()..node.end_byte()]),
+                color,
+                '`',
+                row_width,
+                indent_level,
+            )?;
+        }
+    } else {
+        let color = if node.is_missing() || !source_code.is_utf8() {
```

Comment:
same as earlier, non-utf8 text isnt inherently erroneous, and missing might be better modeled as a "warning"